internal/curator/storageclass/storageclass.go

```diff
@@ -62,74 +62,92 @@ func Get(id core.StorageClass) Class {
-	return len(tract.Hosts) >= 1
+func (r repl) Has(tract *fb.TractF) bool {
+	return tract.HostsLength() >= 1
 }
 
-func (r repl) Clear(tract *pb.Tract) {
+func (r repl) Clear(tract *fb.Tract) {
 	tract.Hosts = nil
 }
 
-func (r repl) Set(tract *pb.Tract, key []byte) core.Error {
+func (r repl) Set(tract *fb.Tract, cid core.RSChunkID) core.Error {
 	return core.ErrInvalidArgument
 }
 
 func (r repl) RSParams() (int, int) {
 	return -1, -1
 }
 
-func (r repl) GetRS(tract *pb.Tract) []byte {
-	return nil
+func (r repl) GetRS(tract *fb.TractF) core.RSChunkID {
+	return core.RSChunkID{}
 }
 
 // implements Class
 type rs struct {
-	id    core.StorageClass // the storage class ID
-	n, m  int               // RS parameters
-	field int               // index of RsNMChunk field in pb.Tract
+	id     core.StorageClass // the storage class ID
+	n, m   int               // RS parameters
+	field  int               // index of RsNMChunk field in fb.Tract
+	method int               // index of RsNMChunk method in fb.TractF
 }
 
 func makeRS(id core.StorageClass, n, m int) rs {
 	name := fmt.Sprintf("Rs%d%dChunk", n, m)
-	sf, ok := reflect.TypeOf(pb.Tract{}).FieldByName(name)
+	sf, ok := reflect.TypeOf(fb.Tract{}).FieldByName(name)
 	if !ok || len(sf.Index) != 1 {
-		panic(fmt.Sprintf("missing %s field in Tract proto message, or len(Index) != 1", name))
+		panic(fmt.Sprintf("missing %s field in Tract struct, or len(Index) != 1", name))
+	}
+	sff, ok := reflect.TypeOf(&fb.TractF{}).MethodByName(name)
+	if !ok {
+		panic(fmt.Sprintf("missing %s method in TractF struct", name))
 	}
-	return rs{id, n, m, sf.Index[0]}
+	return rs{id, n, m, sf.Index[0], sff.Index}
 }
 
 func (r rs) ID() core.StorageClass {
 	return r.id
 }
 
-func (r rs) val(tract *pb.Tract) reflect.Value {
+func (r rs) val(tract *fb.Tract) reflect.Value {
 	return reflect.ValueOf(tract).Elem().Field(r.field)
 }
 
-func (r rs) Has(tract *pb.Tract) bool {
-	return !r.val(tract).IsNil()
+func (r rs) valfb(tract *fb.TractF) reflect.Value {
+	return reflect.ValueOf(tract).Method(r.method)
 }
 
-func (r rs) Clear(tract *pb.Tract) {
-	r.val(tract).SetBytes(nil)
+func (r rs) Has(tract *fb.TractF) bool {
+	f := r.valfb(tract).Interface().(func(*fb.TractIDF) *fb.TractIDF)
+	var t fb.TractIDF
+	return f(&t) != nil
 }
 
-func (r rs) Set(tract *pb.Tract, key []byte) core.Error {
+func (r rs) Clear(tract *fb.Tract) {
+	var none *core.TractID
+	r.val(tract).Set(reflect.ValueOf(none))
+}
+
+func (r rs) Set(tract *fb.Tract, cid core.RSChunkID) core.Error {
 	v := r.val(tract)
 	if !v.IsNil() {
 		return core.ErrConflictingState
 	}
-	v.SetBytes(key)
+	tid := cid.ToTractID()
+	v.Set(reflect.ValueOf(&tid))
 	return core.NoError
 }
 
 func (r rs) RSParams() (int, int) {
 	return r.n, r.m
 }
 
-func (r rs) GetRS(tract *pb.Tract) []byte {
-	return r.val(tract).Bytes()
+func (r rs) GetRS(tract *fb.TractF) core.RSChunkID {
+	f := r.valfb(tract).Interface().(func(*fb.TractIDF) *fb.TractIDF)
```

Comment:
Is this part of the flatbuffer interface? What is the idea here and where can I read about it?

Reply:
I guess the comments aren't too detailed... this whole file is just a wrapper around some useful bits of reflection. The problem it's solving is: let's say you have a blob with storage class `RS_8_3`. How do you go from there to things like `tractf.Rs83Chunk()` and `tract.Rs83Chunk = myNewChunk`? (Without putting a big switch statement everywhere, which would make it hard to evolve the code.) The solution is to turn the storage class enum into a value that implements `storageclass.Class`, and call methods on that value to do the things you need. And we can kinda shoehorn `REPLICATED` into this scheme too, so do that.

So you call `Get` first to look up the `Class` for a given enum value, then call methods on that. Or loop over all of them in `All` or `AllRS`.